### DISCUSSION TITLE
Copying/moving with a different principle

### DISCUSSION THREAD

**Discussion by og900aero** [gokcehan/lf#1553] [General]
I want to implement the copy and move command in reverse. In other words, we do not select with space what we want to perform an operation with, then decide with a button whether to copy (y) or cut (d), and then paste with a (p) button, but we select with space what we want to perform an operation with, then copy and paste with a button (p), and cut and paste with a button (P) (like vifm does). I think this would be a simpler and faster solution. I ran into a problem, because it prints an error on the select command because an argument is missing. I wrote the following code:

```
cmd toggle_new &{{
	echo $f >> ~/.local/share/lf/select_files
}}

cmd reset_select_files &{{
	truncate -s 0 ~/.local/share/lf/select_files 
}}

cmd copy_new &{{
	cat ~/.local/share/lf/select_files | while read -r line
	do
        lf -remote "send select"
    done
	# lf -remote "send copy"
	# lf -remote "send paste"
	# lf -remote "send clear"
	# lf -remote "send unselect"
	# lf -remote "send reset_select_files"
}}

cmd move_new &{{
	cat ~/.local/share/lf/select_files | while read -r line
	do
        lf -remote "send select"
    done
	lf -remote "send cut"
	lf -remote "send paste"
	lf -remote "send clear"
	lf -remote "send unselect"
	lf -remote "send reset_select_files"
}}

map <space> :toggle_new; toggle; down
map p :copy_new
map P :move_new
```
I get error message: "select: requires an argument"

What is the problem? The goal would be for the code to work even if only one lf window is open, and for it to work even if I copy it from one lf window to another lf window.

**Comment by DusanLesan** [comment]
The problem is that select requires a path of file/dir to select. So if you want to select all files in a new dir after copy, you would need to pass valid paths of all files. Something like: 
```
	cat ~/.local/share/lf/select_files | while read -r line
	do
		lf -remote "send select $new_dir/$(basename "$line")"
	done
```

**Reply by og900aero** [reply]
No. The way it would work is that I use space to select which files I want to operate on, I saved these files with the full path to ~/.local/share/lf/select_files. Then, when I press p or P in the other window, these files it has stored are also selected in the new window, so that if I press the copy or cut command afterwards, it can interpret them.

**Reply by DusanLesan** [reply]
Still, my point is the same. You must send argument to select. Also actually, if I am not mistaken, it does not make sense to select in a loop since select just moves focus to that file so you will just end up moving current position to next item.

**Reply by joelim-work** [reply]
The `select` command focuses the cursor on a file, which is a different concept. If you want to add files to the selection list, you need to use the `toggle` command. 

Something like this:

```shell
lf -remote "send toggle $(tr '\n', ' ' < ~/.local/share/lf/select_files)"
```

**Reply by og900aero** [reply]
I try lf -remote "send toggle $(tr '\n', ' ' < ~/.local/share/lf/select_files)", but I get error message:

toggle: lstat /home/Data/fu.mp4 no such file or directory

It looks like it is wrong to separate the files from the file, because I did not select the fu.mp4 file, but there was "ving tsun kung fu.mp4"

**Reply by joelim-work** [reply]
Yeah I couldn't be bothered handling filenames with spaces in them (should be possible of course), the main purpose was to show that you should be using `toggle`, not `select`.

**Reply by og900aero** [reply]
Work with this code when I use 2 lf:

```
cmd copy_new &{{
	cat ~/.local/share/lf/select_files | while read -r line
	do
             lf -remote "send toggle \"$line\""
        done
	lf -remote "send $id copy"
	lf -remote "send $id paste"
	lf -remote "send clear"
	lf -remote "send unselect"
	lf -remote "send reset_select_files"
}}
```

But doesnt work, if I use only one lf, because the toggle command turn off selected files :D So, there should be a check to see if there are selections in the active lf window. If there aren't, you can run the toggle line, if there are, you can't.

**Reply by DusanLesan** [reply]
You can maybe just use `cp` and `mv` to handle files in select_files and then only reset/clear...

**Reply by DusanLesan** [reply]
Actually one of solutions I gave you for this issue last year still works fine for me (with slight changes): https://github.com/gokcehan/lf/issues/1037#issuecomment-1337899929

<details><summary>Almost the same code like a year ago</summary>
<p>

```
cmd signalMoveOrCopy ${{
	:> ~/.local/share/lf/custom_file_list
	pgrep -f "^lf -print-last-dir" | xargs -I {} lf -remote "send {} saveSelected"
	lf -remote "send $id processList $1"
}}

cmd processList ${{
	for file in $(< ~/.local/share/lf/custom_file_list); do
		[[ $1 == "copy" ]] && cp -r "$file" . || mv -r "$file" .
	done
	lf -remote "send unselect"
}}

cmd saveSelected ${{
	[ -n "$fs" ] && echo "$fs" >> ~/.local/share/lf/custom_file_list
}}

map P :signalMoveOrCopy move
map p :signalMoveOrCopy copy
```

</p>
</details>

**Comment by og900aero** [comment]
> Actually one of solutions I gave you for this issue last year still works fine for me (with slight changes): [#1037 (comment)](https://github.com/gokcehan/lf/issues/1037#issuecomment-1337899929)
> 
> Almost the same code like a year ago

Yes, but I want use the built-in copy and move, because the progress bar. advcp, advmv, rsync not alternative, doesnt work fine with progress bar. 
Checking with grep/pgrep is not good, because there are cases where an lf is open, in which I start a text editor (micro), select it in the other lf, and then in the same lf I want to paste it back into another directory , it detects with pgrep that there are two lfs, and then uses the toggle switch, even though it shouldn't, because I want to use it in the same lf as the one I selected. I need some kind of query command that I can use to query a current lf to see if it is currently selected files or directory.

**Comment by DusanLesan** [comment]
> Yes, but I want use the built-in copy and move, because the progress bar. advcp, advmv, rsync not alternative, doesnt work fine with progress bar. Checking with grep/pgrep is not good, because there are cases where an lf is open, in which I start a text editor (micro), select it in the other lf, and then in the same lf I want to paste it back into another directory , it detects with pgrep that there are two lfs, and then uses the toggle switch, even though it shouldn't, because I want to use it in the same lf as the one I selected. I need some kind of query command that I can use to query a current lf to see if it is currently selected files or directory.


Makes sense. How about modifying that code a little to suit your requirements. Like this:
```
cmd signalMoveOrCopy ${{
	:> ~/.local/share/lf/custom_file_list
	pgrep -f "^lf -print-last-dir" | xargs -I {} lf -remote "send {} saveSelected"
	lf -remote "send $id processList $1"
}}

cmd processList ${{
	lf -remote "send unselect"
	for file in $(< ~/.local/share/lf/custom_file_list); do
		lf -remote "send $id toggle $file"
	done
	lf -remote "send $id $1"
	lf -remote "send $id paste"
}}

cmd saveSelected ${{
	[ -n "$fs" ] && echo "$fs" >> ~/.local/share/lf/custom_file_list
}}

map P :signalMoveOrCopy cut
map p :signalMoveOrCopy copy
```

**Reply by og900aero** [reply]
Doesnt work: "paste: no file in copy/cut buffer"

**Reply by DusanLesan** [reply]
Meaning that toggle did not work well. Debug it part by part. Maybe first make sure that all clients pids are determined (last year you had issues with that part). Try making running parts of the script separately

**Reply by DusanLesan** [reply]
You can use tools like notify-send or something similar to print different states or values.

**Reply by og900aero** [reply]
Oh my god! You give me an excellent idea! 

```
cmd processList ${{
	lf -remote "send unselect"
	for file in $(< ~/.local/share/lf/custom_file_list); do
		lf -remote "send $id toggle $file"
	done
	lf -remote "send $id $1"
	lf -remote "send $id paste"
}}
```
The key is the lf -remote "send unselect" at the beginning. You are a genius!
Now I modified my code, and now working with one lf and more lf too, like a vifm, yuhuuuuu! :D

The full code became below:

```
cmd toggle_new &{{
	echo $f >> ~/.local/share/lf/select_files
}}

cmd reset_select_files &{{
	truncate -s 0 ~/.local/share/lf/select_files 
}}

cmd signalMoveOrCopy &{{
	lf -remote "send unselect"
	cat ~/.local/share/lf/select_files | while read -r line
	do
             lf -remote "send toggle \"$line\""
        done
	lf -remote "send $id $1"
	lf -remote "send $id paste"
	lf -remote "send clear"
	lf -remote "send unselect"
	lf -remote "send reset_select_files"
}}

map q :reset_select_files; quit
map <space> :toggle_new; toggle; down
map p :signalMoveOrCopy copy
map P :signalMoveOrCopy cut
```

**Reply by joelim-work** [reply]
This seems like a useful idea, I have added something similar to the wiki: https://github.com/gokcehan/lf/wiki/Tips#alternate-paste-workflow-select-then-movecopy

**Reply by DusanLesan** [reply]
I did think about select sync, but I did now have time to think about syncing new clients when they are opened. Maybe it is good to consider to have cross-client sync as built in option (all clients use the same $fs stored in server or something like that)

**Reply by joelim-work** [reply]
In theory it should be possible to sync new clients as well, just write out the selection list to a file, and `toggle` the contents during start up. I was very torn between providing the bare minimum vs. making the example snippet as comprehensive as possible.

I didn't test it extremely thoroughly, feel free to play around with it and edit it as you please.

**Reply by DusanLesan** [reply]
Maybe it is not necessary to unselect and toggle all again if we sync on start. 
If we have every instance of client send the file currently toggled to all others and if we have all new ones pick up all selections:

Like:
```
cmd select-sync &{{
	[ -z "$1" ] && lf -remote "send unselect"
	if [ -n "$fs" ]; then
		files=$(echo "$fs" | tr '\n' '\0' | xargs -0 printf ' %q')
		lf -remote "send $1 toggle $files"
	fi
}}

cmd get-data &{{
	for pid in $(prep -f "^lf -print-last-dir"); do
		if [ "$pid" != "$id" ]; then
			lf -remote "send $pid select-sync $id"
			break
		fi
	done
}}

cmd toggle-sync &{{
	lf -remote "send toggle $f"
}}

get-data

map <space> :toggle-sync; down;
map u :unselect; select-sync
map v :invert; select-sync
```
So toggle will be done for all instances, and on start `get-data` will call one other existing instance to update itself. Of-course this is only a example and will probably not work for og900aero because pgrep used here searches for -print-last-dir to avoid hitting server but is probably not used by all users

**Reply by joelim-work** [reply]
Thanks for the feedback. The idea of using `$id` is quite useful and I found a way to use it so an instance doesn't need to refresh itself as mentioned by @og900aero. I updated the wiki example.

**Comment by og900aero** [comment]
> This seems like a useful idea, I have added something similar to the wiki: https://github.com/gokcehan/lf/wiki/Tips#alternate-paste-workflow-select-then-movecopy

The problem with this solution is that, while it works well, the selection flashes every time you select it, and it slows things down a bit. However, writing the selection to a file is fast enough and does not affect the UI.

**Comment by og900aero** [comment]
DusanLesan and joelim-work thank u for your hard work and best wishis for christmas event! :)

**Comment by og900aero** [comment]
Meanwhile, I realized that it is even more complicated. I still had to solve it so that the invert command also works with this method, pay attention to the space so that if you delete the selection, you also delete the link from selected_files, and even there the problem is the push :glob-select<space>*.
Do you think there is no chance that this kind of copy/cut/paste solution will be included in the program at an official level?

**Reply by joelim-work** [reply]
I did not consider `glob-select`/`glob-unselect`. TBH I didn't even know those commands existed (even as a contributor there are things I don't know about 🙂).

What you're actually asking here is the ability to sync the selection between clients (just like the cup/copy list), and this is a completely separate topic from the alternate copy/move pasting model.

The current design is that each client can have their own selection list, however I don't know the reason behind this decision if any, I guess it would be possible to change it so that there is only one selection list that is shared among all instances, but that would be a breaking change and probably requires more discussion before doing so. In the meantime you will have to use a file like `~/.local/share/lf/select` as shown in my example as a hacky way to achieve this.

**Reply by og900aero** [reply]
Unfortunately, the example you wrote didn't work for me in all cases, so I rewrote it a bit, and now it works for 1 or more copies of copying and moving by simply selecting one by one (with space), and deselecting also works. It also works using the invert command. There is only a problem with globe-select, but I also wrote a ticket about it: https://github.com/gokcehan/lf/discussions/1560
